### PR TITLE
fix: pending transaction ordering

### LINKED ISSUE #12286: Two transactions from the same sender exist in `PendingPool::independent_transactions`, causing "nonce too high" warnings
### Describe the bug

In some circumstances, `PendingPool::independent_transactions` might contain two transactions from the same sender.
This causes `BestTransactions::new(..)` to emit duplicated transactions, and hence, cause "nonce too high" warning.

Also, in a brief moment, `PendingPool` does contain a nonce gap.

### Steps to reproduce

Add the following test to `crates/transaction-pool/src/pool/txpool.rs` then run it.

```rust
#[test]
fn one_sender_two_independent_transactions() {
    let mut on_chain_balance = U256::from(4_999); // only enough for 4 txs
    let mut on_chain_nonce = 40;
    let mut f = MockTransactionFactory::default();
    let mut pool = TxPool::mock();
    let mut submitted_txs = Vec::new();

    // We use a "template" because we want all txs to have the same sender.
    let template =
        MockTransaction::eip1559().inc_price().inc_limit().with_value(U256::from(1_001));

    // Add 8 txs. Because the balance is only sufficient for 4, so the last 4 will be
    // Queued.
    for tx_nonce in 40..48 {
        let tx = f.validated(template.clone().with_nonce(tx_nonce).rng_hash());
        submitted_txs.push(*tx.id());
        pool.add_transaction(tx, on_chain_balance, on_chain_nonce).unwrap();
    }

    // A block is mined with two txs (so nonce is changed from 40 to 42).
    // Now the balance gets so high that it's enough to execute alltxs.
    on_chain_balance = U256::from(999_999);
    on_chain_nonce = 42;
    pool.remove_transaction(&submitted_txs[0]);
    pool.remove_transaction(&submitted_txs[1]);

    // Add 4 txs.
    for tx_nonce in 48..52 {
        pool.add_transaction(
            f.validated(template.clone().with_nonce(tx_nonce).rng_hash()),
            on_chain_balance,
            on_chain_nonce,
        )
        .unwrap();
    }

    let best_txs: Vec<_> = pool.pending().best().map(|tx| *tx.id()).collect();
    assert_eq!(best_txs.len(), 10); // 8 - 2 + 4 = 10

    // Another useful assertion.
    // assert_eq!(pool.pending().num_independent_transactions(), 1);
}
```

### Node logs

_No response_

### Platform(s)

Mac (Apple Silicon)

### What version/commit are you on?

v1.1.0

### What database version are you on?

N/A

### Which chain / network are you on?

N/A

### What type of node are you running?

Archive (default)

### What prune config do you use, if any?

_No response_

### If you've built Reth from source, provide the full command you used

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### LINKED ISSUE #12340: In `BestTransactions`, some updates are not marked as independent although they should be
### Describe the bug

In `BestTransactions`, although the call `self.add_new_transactions()` adopts incoming transactions from `self.new_transaction_receiver`, it does _not_ mark a valid tx as independent.

The following `if` statement is _incorrect_:

```rust
// crates/transaction-pool/src/pool/best.rs
if self.ancestor(&tx_id).is_none() {
    self.independent.insert(pending_tx.clone());
}
```

Let's read the implementation of `.ancestor`:

```rust
pub(crate) fn ancestor(&self, id: &TransactionId) -> Option<&PendingTransaction<T>> {
    self.all.get(&id.unchecked_ancestor()?)
}
```

So, the `if` statement above can be understood as:

_If `self.all` does **not** contain the parent (a.k.a. direct ancestor) of `tx_id`, then `tx_id` is considered as independent._

However, `self.all` is an "append-only" field (there are no `self.all.remove(...)` calls in the code base). It means that, once a tx exists in `self.all`, it will be there forever, even after a tx is emitted by `.next()`.

**Event sequence**
1. A `BestTransactions` instance is initialized with txs 42, 43, 44
2. `.next()` is called 3 times, emitting 42, 43, 44
3. `.try_recv` receives a new tx, 45, however, this tx is _not_ marked as independent
4. `.next()` returns `None`, while it should returns `Some(45)` instead

**Suggested solution**

Either
we let `.next()` removes the emitted tx from `self.all`,

or
fix the `if` statement in `fn add_new_transactions(&mut self)`.

### Steps to reproduce

Add this test to `crates/transaction-pool/src/pool/best.rs` and run it:

```rust
#[test]
fn test_eligible_updates_not_promoted() {
    let mut pool = PendingPool::new(MockOrdering::default());
    let mut f = MockTransactionFactory::default();

    let num_senders = 10;

    let first_txs: Vec<_> = (0..num_senders) //
        .map(|_| MockTransaction::eip1559())
        .collect();
    let second_txs: Vec<_> =
        first_txs.iter().map(|tx| tx.clone().rng_hash().inc_nonce()).collect();

    for tx in first_txs {
        let valid_tx = f.validated(tx);
        pool.add_transaction(Arc::new(valid_tx), 0);
    }

    let mut best = pool.best();

    for _ in 0..num_senders {
        if let Some(tx) = best.next() {
            println!("{:?}", tx.transaction_id);
        } else {
            panic!("cannot read one of first_txs");
        }
    }

    for tx in second_txs {
        let valid_tx = f.validated(tx);
        pool.add_transaction(Arc::new(valid_tx), 0);
    }

    for _ in 0..num_senders {
        if let Some(tx) = best.next() {
            println!("{:?}", tx.transaction_id);
        } else {
            panic!("cannot read one of second_txs");
        }
    }
}
```

### Node logs

_No response_

### Platform(s)

Mac (Apple Silicon)

### What version/commit are you on?

v1.1.0

### What database version are you on?

N/A

### Which chain / network are you on?

N/A

### What type of node are you running?

Archive (default)

### What prune config do you use, if any?

_No response_

### If you've built Reth from source, provide the full command you used

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by klkvr**
Closes https://github.com/paradigmxyz/reth/issues/12340
Closes https://github.com/paradigmxyz/reth/issues/12286

Updates the way transaction adding is handled in `Pending` pool.
- `add_transaction`  now correctly handles inserted transactions and does not make assumptions on their order. This is done by stroing a mapping sender -> higher/lower nonce instead of a simple set of highest/lowest transactions
- whenever we update `BestTransactions` we now pre-sort the transactions vec by nonce to make sure that there will be no nonce gaps. I didn't apply same approach as with pending pool here because we insert pretty rary and need ability to get the independet tx with highes priority which is hard to do in a nice way with a hashmap

### CODE REVIEW

File: crates/transaction-pool/src/pool/pending.rs [hunk 1]
```diff
@@ -320,19 +320,26 @@ impl<T: TransactionOrdering> PendingPool<T> {
...
-        // mark the next as independent if it exists
-        if let Some(unlocked) = self.get(&id.descendant()) {
-            self.independent_transactions.insert(unlocked.clone());
+        if let Some(lowest) = self.independent_transactions.get(&id.sender) {
+            if lowest.transaction.nonce() == id.nonce {
+                self.independent_transactions.remove(&id.sender);
+                // mark the next as independent if it exists
+                if let Some(unlocked) = self.get(&id.descendant()) {
+                    self.independent_transactions.insert(id.sender, unlocked.clone());
+                }
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/pending.rs:329]: this is the proper logic now

File: crates/transaction-pool/src/pool/pending.rs [hunk 2]
```diff
@@ -255,17 +252,20 @@ impl<T: TransactionOrdering> PendingPool<T> {
...
-            // the transaction already has an ancestor, so we only need to ensure that the
-            // highest nonces set actually contains the highest nonce for that sender
-            self.highest_nonces.remove(ancestor);
+        if let Some(highest) = self.highest_nonces.get_mut(&tx.transaction.sender_id()) {
+            if highest.transaction.nonce() < tx.transaction.nonce() {
+                *highest = tx.clone();
+            }
+        } else {
+            self.highest_nonces.insert(tx.transaction.sender_id(), tx.clone());
+        }
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/pending.rs:261]: can we solve this with the entry api?

File: crates/transaction-pool/src/pool/pending.rs [hunk 3]
```diff
@@ -255,17 +252,20 @@ impl<T: TransactionOrdering> PendingPool<T> {
...
+        if let Some(lowest) = self.independent_transactions.get_mut(&tx.transaction.sender_id()) {
+            if lowest.transaction.nonce() > tx.transaction.nonce() {
+                *lowest = tx.clone();
+            }
         } else {
-            // If there's __no__ ancestor in the pool, then this transaction is independent, this is
-            // guaranteed because this pool is gapless.
-            self.independent_transactions.insert(tx.clone());
+            self.independent_transactions.insert(tx.transaction.sender_id(), tx.clone());
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/pending.rs:267]: looks like this can also be `match entry`?

I'd find that easier to follow

**Review by mattsse — CHANGES_REQUESTED** [review]
style nits

File: crates/transaction-pool/src/pool/best.rs [hunk 4]
```diff
@@ -131,7 +131,12 @@ impl<T: TransactionOrdering> BestTransactions<T> {
     /// Checks for new transactions that have come into the `PendingPool` after this iterator was
     /// created and inserts them
     fn add_new_transactions(&mut self) {
+        let mut new_txs = Vec::new();
         while let Some(pending_tx) = self.try_recv() {
+            new_txs.push(pending_tx);
+        }
+        new_txs.sort_by_key(|tx| tx.transaction.nonce());
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/best.rs:138]: this looks a bit weird, why can we no longer process txs one by one?

how many txs are buffered in `new_txs` is arbitrary, so doesn't make sense that we have to sort this arbitrary chunk

File: crates/transaction-pool/src/pool/txpool.rs [hunk 5]
```diff
@@ -1676,29 +1676,14 @@ impl<T: PoolTransaction> AllTransactions<T> {
...
-            } else {
-                // The transaction was added above so the _inclusive_ descendants iterator
-                // returns at least 1 tx.
-                let (id, tx) = descendants.peek().expect("includes >= 1");
-                if id.nonce < inserted_tx_id.nonce {
-                    !tx.state.is_pending()
-                } else {
-                    true
-                }
-            };
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/txpool.rs:1698]: hmm, this now means that we're always hitting:

https://github.com/paradigmxyz/reth/blob/22d740cad21eb677f610c4cc16f9bd7e7c369c0a/crates/transaction-pool/src/pool/txpool.rs#L1714-L1714

this feels off

File: crates/transaction-pool/src/pool/best.rs (hunk 4 above)
Comment (klkvr) [reply crates/transaction-pool/src/pool/best.rs]: updated `BestTransactions`, it now holds the following:
https://github.com/paradigmxyz/reth/blob/311a10d7b950c11aa3a970db11fdb4c0ac06d63b/crates/transaction-pool/src/pool/best.rs#L77-L83

when inserting, we make sure to replace independent tx if the nonce of tx being added is lower than the lowest we know for the given sender

that way we still don't make assumptions on order of transactions and don't have to sort everything

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**